libfreerdp/core/gateway/arm.c

```diff
@@ -0,0 +1,392 @@
+{
+	int status = -1;
+	wStream* s =
+	    arm_build_http_request(arm, method, TransferEncodingIdentity, content_type, content_length);
+
+	if (!s)
+		return FALSE;
+
+	const size_t sz = Stream_Length(s);
+
+	if (sz <= INT_MAX)
+		status = freerdp_tls_write_all(tls, Stream_Buffer(s), (int)sz);
+
+	Stream_Free(s, TRUE);
+	if (status >= 0 && content_length > 0 && data)
+		status = freerdp_tls_write_all(tls, (const BYTE*)data, content_length);
+
+	return (status >= 0);
+}
+
+static void arm_free(rdpArm* arm)
+{
+	if (!arm)
+		return;
+
+	freerdp_tls_free(arm->tls);
+	http_context_free(arm->http);
+
+	free(arm);
+}
+
+static char* arm_create_request_json(rdpArm* arm)
+{
+	char* lbi = NULL;
+	cJSON* json = NULL;
+	char* message = NULL;
+	json = cJSON_CreateObject();
+	if (!json)
+		goto arm_create_cleanup;
+	cJSON_AddStringToObject(
+	    json, "application",
+	    freerdp_settings_get_string(arm->settings, FreeRDP_RemoteApplicationProgram));
+
+	lbi = calloc(freerdp_settings_get_uint32(arm->settings, FreeRDP_LoadBalanceInfoLength) + 1,
+	             sizeof(char));
+	if (!lbi)
+		goto arm_create_cleanup;
+
+	memcpy(lbi, freerdp_settings_get_pointer(arm->settings, FreeRDP_LoadBalanceInfo),
+	       freerdp_settings_get_uint32(arm->settings, FreeRDP_LoadBalanceInfoLength));
+
+	cJSON_AddStringToObject(json, "loadBalanceInfo", lbi);
+	free(lbi);
+	lbi = NULL;
+
+	cJSON_AddNullToObject(json, "LogonToken");
+	cJSON_AddNullToObject(json, "gatewayLoadBalancerToken");
+
+	message = cJSON_PrintUnformatted(json);
+arm_create_cleanup:
+	if (json)
+		cJSON_Delete(json);
+	free(lbi);
+	return message;
+}
+
+static BOOL arm_fill_gateway_parameters(rdpArm* arm, const char* message)
+{
+	WINPR_ASSERT(arm);
+	WINPR_ASSERT(arm->settings);
+	WINPR_ASSERT(message);
+	cJSON* json = cJSON_Parse(message);
+	BOOL status = FALSE;
+	if (!json)
+		return FALSE;
+	else
+	{
+		const cJSON* gwurl = cJSON_GetObjectItemCaseSensitive(json, "gatewayLocation");
+		if (cJSON_IsString(gwurl) && (gwurl->valuestring != NULL))
+		{
+			WLog_DBG(TAG, "extracted target url %s", gwurl->valuestring);
+			if (!freerdp_settings_set_string(arm->settings, FreeRDP_GatewayUrl, gwurl->valuestring))
+				status = FALSE;
+			else
+				status = TRUE;
+		}
+		cJSON_Delete(json);
+	}
+	return status;
+}
+
+#endif
+
+BOOL arm_resolve_endpoint(rdpContext* context, DWORD timeout)
+{
+#ifndef WITH_CJSON
+	WLog_ERR(TAG, "arm gateway support not compiled in");
+	return FALSE;
+#else
+
+	if (!context)
+		return FALSE;
+
+	if (!context->settings)
+		return FALSE;
+
+	if ((freerdp_settings_get_uint32(context->settings, FreeRDP_LoadBalanceInfoLength) == 0) ||
+	    (freerdp_settings_get_string(context->settings, FreeRDP_RemoteApplicationProgram) == NULL))
+	{
+		WLog_ERR(TAG, "loadBalanceInfo and RemoteApplicationProgram needed");
+		return FALSE;
+	}
+
+	rdpArm* arm = (rdpArm*)calloc(1, sizeof(rdpArm));
+	char* message = NULL;
```

Comment:
if calloc fails the function still returns true

Reply:
no more